File: Python/textbook_track/chapter10/ba10i.py
# Implement Viterbi Learning
import math
import sys

from collections import namedtuple

EPSILON = 1e-10
DIGITS = 3

HMM = namedtuple('HMM', ['alphabet', 'states', 'transition', 'emission'])


class Label:
    def __init__(self, labels):
        self._labels = labels
        self._label_indices = {item: ix for ix, item in enumerate(labels)}
        self._size = len(labels)

    def __getitem__(self, label):
        return self._label_indices[label]

    def __repr__(self):
        return ', '.join(map(str, self._labels))

    def __iter__(self):
        yield from self._labels

    @property
    def labels(self):
        return self._labels

    @property
    def size(self):
        return self._size


class ProbabilityMatrix:
    def __init__(self, row_labels, column_labels, probabilities=None):
        self._row_labels = row_labels
        self._column_labels = column_labels
        self._probabilities = self._initialize_matrix(probabilities)

    def __getitem__(self, label_pair):
        row_label, col_label = label_pair
        return self._probabilities[self._row_labels[row_label]][self._column_labels[col_label]]

    def __setitem__(self, label_pair, value):
        row_label, col_label = label_pair
        self._probabilities[self._row_labels[row_label]][self._column_labels[col_label]] = value

    def __repr__(self):
        lines = ['\t' + '\t'.join(self._column_labels)]
        for label in self._row_labels:
            probs = self._probabilities[self._row_labels[label]]
            row = '\t'.join(str(round(p, DIGITS)) for p in probs)
            lines.append(f'{label}\t{row}')
        return '\n'.join(lines)

    def _initialize_matrix(self, probabilities):
        if probabilities is not None:
            return probabilities

        return [[EPSILON] * self.nr_cols for _ in range(self.nr_rows)]

    def rowsum(self, label):
        return sum(self._probabilities[self._row_labels[label]])

    @property
    def nr_rows(self):
        return self._row_labels.size

    @property
    def nr_cols(self):
        return self._column_labels.size


def read_matrix(reader, k):
    _ = next(reader)
    matrix = []
    for _ in range(k):
        line = next(reader)
        probabilities = [float(item) for item in line.split()[1:]]
        matrix.append(probabilities)
    return matrix


def read_data(reader):
    n = int(next(reader))
    _ = next(reader)
    string = next(reader).rstrip()
    _ = next(reader)
    alphabet = Label(next(reader).split())
    _ = next(reader)
    states = Label(next(reader).split())
    _ = next(reader)
    transition = ProbabilityMatrix(states, states, read_matrix(reader, states.size))
    _ = next(reader)
    emission = ProbabilityMatrix(states, alphabet, read_matrix(reader, states.size))
    hmm = HMM(alphabet=alphabet, states=states, transition=transition, emission=emission)
    return string, hmm, n


def calc_argmax(sequence):
    return max(enumerate(sequence), key=lambda x: x[1])


def estimate_transition_probabilities(hidden_path, states):
    transition = ProbabilityMatrix(states, states)
    for state_from, state_to in zip(hidden_path, hidden_path[1:]):
        transition[state_from, state_to] += 1

    for s1 in states:
        rowsum = transition.rowsum(s1)
        for s2 in states:
            transition[s1, s2] /= rowsum
    return transition


def estimate_emission_probabilities(emitted_string, hidden_path, alphabet, states):
    emission = ProbabilityMatrix(states, alphabet)
    for state, char in zip(hidden_path, emitted_string):
        emission[state, char] += 1

    for state in states:
        rowsum = emission.rowsum(state)
        for letter in alphabet:
            emission[state, letter] /= rowsum
    return emission


def estimate_hmm_parameters(emitted_string, hidden_path, alphabet, states):
    transition = estimate_transition_probabilities(hidden_path, states)
    emission = estimate_emission_probabilities(emitted_string, hidden_path, alphabet, states)
    return HMM(alphabet, states, transition, emission)


def reconstruct_optimal_hidden_path(backtrack, nr_cols, maximal_score_index, states):
    hidden_path = [states.labels[maximal_score_index]]
    jy = nr_cols
    while jy > 0:
        jy -= 1
        maximal_score_index = backtrack[maximal_score_index][jy]
        hidden_path.append(states.labels[maximal_score_index])
    return ''.join(hidden_path[::-1])


def calc_most_likely_hidden_path(hmm, emitted_string):
    _, states, transition, emission = hmm
    nr_cols = len(emitted_string)
    logscore_matrix = [[0.0] * nr_cols for _ in range(states.size)]
    backtrack = [[0] * (nr_cols - 1) for _ in range(states.size)]
    char = emitted_string[0]
    for ix, state in enumerate(states):
        logscore_matrix[ix][0] = math.log(1 / states.size) + math.log(emission[state, char])

    for jy, char in enumerate(emitted_string[1:], start=1):
        for ix, state in enumerate(states):
            emission_logprob = math.log(emission[state, char])
            logscores = [logscore_matrix[k][jy-1] + math.log(transition[previous_state, state]) + emission_logprob
                         for k, previous_state in enumerate(states)]
            index, maximum_sum_weight = calc_argmax(logscores)
            logscore_matrix[ix][jy] = maximum_sum_weight
            backtrack[ix][jy-1] = index
    maximal_score_index, _ = calc_argmax(map(lambda x: x[-1], logscore_matrix))
    return reconstruct_optimal_hidden_path(backtrack, nr_cols-1, maximal_score_index, states)


def run_Viterbi_learning(hmm, emitted_string, n):
    for _ in range(n):
        hidden_path = calc_most_likely_hidden_path(hmm, emitted_string)
        hmm = estimate_hmm_parameters(emitted_string, hidden_path, hmm.alphabet, hmm.states)
    return hmm


def main():
    reader = sys.stdin
    emitted_string, hmm, n = read_data(reader)
    result = run_Viterbi_learning(hmm, emitted_string, n)
    print(result.transition)
    print('--------')
    print(result.emission)


if __name__ == '__main__':
    main()

File: Scala/src/main/scala/TextbookTrack/Chapter10/BA10I.scala
package TextbookTrack.Chapter10

object BA10I {
  import scala.annotation.tailrec

  private type Matrix = Array[Array[Double]]
  private val Digits: Int = 3
  private val Epsilon: Double = 1e-10

  case class Label(labels: Vector[Char]) {
    private val labelIndices: Map[Char, Int] = labels.iterator.zipWithIndex.toMap
    val size: Int = labels.length

    def apply(label: Char): Int = labelIndices(label)

    def foreach(f: Char => Unit): Unit = labels.foreach(f)

    def map[T](f: Char => T): Vector[T] = labels.map(f)

    def zipWithIndex: Iterator[(Char, Int)] = labels.iterator.zipWithIndex
  }

  class ProbabilityMatrix(rowLabels: Label, columnLabels: Label, probs: Option[Matrix] = None) {
    val nrRows: Int = rowLabels.size
    val nrCols: Int = columnLabels.size
    val probabilities: Matrix = initializeMatrix(probs)

    override def toString: String = {
      val header: String = columnLabels.map(_.toString).mkString("\t", "\t", "")
      val lines: Vector[String] =
        rowLabels.map { label =>
          val probs: Array[Double] = probabilities(rowLabels(label))
          s"$label\t${probs.map(round(_, Digits)).mkString("\t")}"
        }
      s"$header\n${lines.mkString("\n")}"
    }

    private def initializeMatrix(probs: Option[Matrix]): Matrix = probs match {
      case Some(matrix) => matrix
      case None => Array.fill(nrRows, nrCols)(Epsilon)
    }

    def apply(row: Char, col: Char): Double = probabilities(rowLabels(row))(columnLabels(col))

    def update(row: Char, col: Char, value: Double): Unit = probabilities(rowLabels(row))(columnLabels(col)) = value

    def rowSum(row: Char): Double = probabilities(rowLabels(row)).sum
  }

  case class HMM(alphabet: Label, states: Label, transition: ProbabilityMatrix, emission: ProbabilityMatrix)

  private def round(number: Double, digits: Int): Double =
    math.rint(math.pow(10, digits) * number) / math.pow(10, digits)

  private def collectLabels(line: String): Vector[Char] =
    line.toVector.filter(_.isLetter)

  private def readMatrix(reader: Iterator[String], k: Int): Array[Array[Double]] = {
    val _: String = reader.next()
    reader
      .take(k)
      .map(_.split("\\s+").drop(1).map(_.toDouble))
      .toArray
  }

  private def readData(reader: Iterator[String]): (String, HMM, Int) = {
    val n: Int = reader.next().toInt
    val _: String = reader.next()
    val string: String = reader.next()
    val _: String = reader.next()
    val alphabet = Label(collectLabels(reader.next()))
    val _: String = reader.next()
    val states = Label(collectLabels(reader.next()))
    val _: String = reader.next()
    val transition = new ProbabilityMatrix(states, states, Some(readMatrix(reader, states.size)))
    val _: String = reader.next()
    val emission = new ProbabilityMatrix(states, alphabet, Some(readMatrix(reader, states.size)))
    val hmm = HMM(alphabet, states, transition, emission)
    (string, hmm, n)
  }

  private def estimateTransitionProbabilities(hiddenPath: String, states: Label): ProbabilityMatrix = {
    val transition = new ProbabilityMatrix(states, states)
    hiddenPath
      .sliding(2)
      .map(_.toList)
      .collect { case List(stateFrom, stateTo) => (stateFrom, stateTo) }
      .foreach { case (stateFrom, stateTo) => transition(stateFrom, stateTo) += 1 }

    for {
      s1 <- states
      rowSum: Double = transition.rowSum(s1)
      s2 <- states
    } transition(s1, s2) /= rowSum

    transition
  }

  private def estimateEmissionProbabilities(emittedString: String,
                                            hiddenPath: String,
                                            alphabet: Label,
                                            states: Label): ProbabilityMatrix = {
    val emission = new ProbabilityMatrix(states, alphabet)
    hiddenPath.lazyZip(emittedString).foreach { case (state, char) => emission(state, char) += 1 }

    for {
      state <- states
      rowSum: Double = emission.rowSum(state)
      letter <- alphabet
    } emission(state, letter) /= rowSum

    emission
  }

  private def estimateHmmParameters(emittedString: String, hiddenPath: String, alphabet: Label, states: Label): HMM = {
    val transition: ProbabilityMatrix = estimateTransitionProbabilities(hiddenPath, states)
    val emission: ProbabilityMatrix = estimateEmissionProbabilities(emittedString, hiddenPath, alphabet, states)
    HMM(alphabet, states, transition, emission)
  }

  private def reconstructOptimalHiddenPath(backtrack: Array[Array[Int]],
                                           nrCols: Int,
                                           maximalScoreIndex: Int,
                                           states: Label): String = {
    @tailrec
    def loop(hiddenPath: List[Char], jy: Int, index: Int): String = {
      if (jy < 0) hiddenPath.mkString
      else {
        val maximalIndex: Int = backtrack(index)(jy)
        loop(states.labels(maximalIndex) :: hiddenPath, jy - 1, maximalIndex)
      }
    }
    loop(List(states.labels(maximalScoreIndex)), nrCols - 1, maximalScoreIndex)
  }

  private def calcMostLikelyHiddenPath(hmm: HMM, emittedString: String): String = {
    val HMM(_, states, transition, emission) = hmm
    val nrCols: Int = emittedString.length
    val logscoreMatrix: Array[Array[Double]] = Array.fill(states.size, nrCols)(0.0)
    val backtrack: Array[Array[Int]] = Array.fill(states.size, nrCols - 1)(0)
    val char: Char = emittedString(0)
    for {
      (state, ix) <- states.zipWithIndex
    } logscoreMatrix(ix)(0) = math.log(1.0 / states.size) + math.log(emission(state, char))

    for {
      (char, jy) <- emittedString.drop(1).zipWithIndex
      (state, ix) <- states.zipWithIndex
    } {
      val emissionLogprob: Double = math.log(emission(state, char))
      val logScores: Iterator[Double] =
        states
          .zipWithIndex
          .map{
            case (previousState, k) =>
              logscoreMatrix(k)(jy) + math.log(transition(previousState, state)) + emissionLogprob
          }
      val (maximumSumWeight, index): (Double, Int) = logScores.zipWithIndex.maxBy{ case (score, _) => score }
      logscoreMatrix(ix)(jy + 1) = maximumSumWeight
      backtrack(ix)(jy) = index
    }

    val (_, maximalScoreIndex): (Double, Int) =
      logscoreMatrix.map(_.last).zipWithIndex.maxBy{ case (score, _) => score }
    reconstructOptimalHiddenPath(backtrack, nrCols - 1, maximalScoreIndex, states)
  }

  def runViterbiLearning(initialHmm: HMM, emittedString: String, n: Int): HMM =
    (0 until n).foldLeft(initialHmm){
      case (hmm, _) =>
        val hiddenPath: String = calcMostLikelyHiddenPath(hmm, emittedString)
        estimateHmmParameters(emittedString, hiddenPath, hmm.alphabet, hmm.states)
    }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val (emittedString, hmm, n): (String, HMM, Int) = readData(reader)
    val result: HMM = runViterbiLearning(hmm, emittedString, n)
    println(result.transition)
    println("--------")
    println(result.emission)
  }
}
